File: scrape_schema.py
import asyncio
import json
import re
from datetime import datetime
from pathlib import Path

from playwright.async_api import async_playwright


BASE_URL = "https://yellowslate.com"
DEFAULT_OUTPUT_DIR = Path("scraped_data")


def get_all_schools(output_dir: Path, cities=None):
    schools = []
    for city_dir in output_dir.iterdir():
        if not city_dir.is_dir() or city_dir.name.startswith("_"):
            continue
        city_slug = city_dir.name
        if cities and city_slug not in cities:
            continue
        
        for school_folder in city_dir.iterdir():
            if not school_folder.is_dir() or school_folder.name.startswith("_"):
                continue
            
            about_file = school_folder / "about.json"
            if not about_file.exists():
                continue
            
            try:
                about_data = json.loads(about_file.read_text())
                if about_data.get("slug"):
                    schools.append({
                        "city": city_slug,
                        "folder": school_folder.name,
                        "slug": about_data["slug"],
                        "name": about_data.get("name", "")
                    })
            except (json.JSONDecodeError, FileNotFoundError):
                continue
    
    return schools


def extract_json_ld(html: str):
    schema_data = []
    pattern = r'<script[^>]*type="application/ld\+json"[^>]*>([\s\S]*?)</script>'
    for match in re.finditer(pattern, html, re.IGNORECASE):
        try:
            data = json.loads(match.group(1))
            schema_data.append(data)
        except json.JSONDecodeError:
            continue
    return schema_data


def parse_breadcrumb(data):
    if data.get("@type") != "BreadcrumbList":
        return None
    items = data.get("itemListElement", [])
    return {
        "items": [
            {"position": item.get("position"), "name": item.get("name"), "url": item.get("item")}
            for item in items
        ]
    }


def parse_localbusiness(data):
    if data.get("@type") != "LocalBusiness":
        return None
    return {
        "name": data.get("name"),
        "image": data.get("image"),
        "url": data.get("url"),
        "telephone": data.get("telephone"),
        "address": data.get("address"),
        "geo": data.get("geo"),
    }


def parse_review_schema(data):
    if data.get("@type") != "Review":
        return None
    item_list = data.get("itemListElement", [])
    reviews = []
    for review in item_list:
        if review.get("@type") != "Review":
            continue
        item_reviewed = review.get("itemReviewed", {})
        author = review.get("author", {})
        rating = review.get("reviewRating", {})
        reviews.append({
            "school_name": item_reviewed.get("name"),
            "school_address": item_reviewed.get("address"),
            "school_url": item_reviewed.get("url"),
            "aggregate_rating": item_reviewed.get("aggregateRating"),
            "author": author.get("name"),
            "date_published": review.get("datePublished"),
            "rating_value": rating.get("ratingValue"),
            "best_rating": rating.get("bestRating"),
            "worst_rating": rating.get("worstRating"),
            "review_body": review.get("reviewBody"),
            "publisher": review.get("publisher", {}).get("name"),
        })
    return {"number_of_items": data.get("numberOfItems"), "reviews": reviews}


def process_schema_data(schema_list):
    result = {"breadcrumbs": [], "local_business": None, "reviews": None}
    
    for schema in schema_list:
        if isinstance(schema, list):
            for item in schema:
                breadcrumb = parse_breadcrumb(item)
                if breadcrumb:
                    result["breadcrumbs"] = breadcrumb["items"]
                
                local_business = parse_localbusiness(item)
                if local_business:
                    result["local_business"] = local_business
                
                review_schema = parse_review_schema(item)
                if review_schema:
                    result["reviews"] = review_schema
        else:
            breadcrumb = parse_breadcrumb(schema)
            if breadcrumb:
                result["breadcrumbs"] = breadcrumb["items"]
            
            local_business = parse_localbusiness(schema)
            if local_business:
                result["local_business"] = local_business
            
            review_schema = parse_review_schema(schema)
            if review_schema:
                result["reviews"] = review_schema
    
    return result


async def scrape_school_schema(browser, school):
    slug = school.get("slug")
    name = school.get("name")
    if not slug:
        return None
    
    detail_url = f"{BASE_URL}{slug}"
    page = await browser.new_page()
    
    try:
        await page.goto(detail_url, wait_until="domcontentloaded", timeout=60000)
        await page.wait_for_timeout(2000)
        
        html = await page.content()
        schema_list = extract_json_ld(html)
        
        if not schema_list:
            print(f"  No JSON-LD schema found for {name}")
            return None
        
        processed = process_schema_data(schema_list)
        processed["url"] = detail_url
        processed["scraped_at"] = datetime.now().isoformat()
        
        return processed
    except Exception as e:
        print(f"  Error scraping {detail_url}: {e}")
        return None
    finally:
        await page.close()


async def run(args):
    output_dir = Path(args.get("output_dir", DEFAULT_OUTPUT_DIR))
    city_filter = args.get("city")
    limit = args.get("limit")
    
    print("Finding schools to process...")
    schools = get_all_schools(output_dir, city_filter)
    print(f"Found {len(schools)} schools")
    
    if limit:
        schools = schools[:limit]
        print(f"Limited to {limit} schools")
    
    async with async_playwright() as playwright:
        browser = await playwright.chromium.launch(headless=True)
        
        processed = 0
        failed = 0
        
        for school in schools:
            print(f"Processing: {school['city']}/{school['folder']}")
            
            schema_data = await scrape_school_schema(browser, school)
            
            if schema_data:
                schema_path = output_dir / school["city"] / school["folder"] / "schema.json"
                schema_path.write_text(json.dumps(schema_data, indent=2, ensure_ascii=False))
                processed += 1
                print(f"  Saved schema.json")
            else:
                failed += 1
                print(f"  Failed to get schema data")
        
        await browser.close()
    
    print(f"\n{'=' * 60}")
    print("SCHEMA SCRAPE SUMMARY")
    print(f"{'=' * 60}")
    print(f"Total schools: {len(schools)}")
    print(f"Processed: {processed}")
    print(f"Failed: {failed}")
    print(f"{'=' * 60}")


def parse_args():
    args = {"output_dir": DEFAULT_OUTPUT_DIR, "city": None, "limit": None}
    
    for arg in __import__("sys").argv[1:]:
        if arg.startswith("--output-dir="):
            args["output_dir"] = arg.split("=")[1]
        elif arg.startswith("--city="):
            cities_str = arg.split("=")[1]
            args["city"] = set(c.strip().lower() for c in cities_str.split(",") if c.strip())
        elif arg.startswith("--limit="):
            try:
                args["limit"] = int(arg.split("=")[1])
            except ValueError:
                pass
    
    return args


if __name__ == "__main__":
    asyncio.run(run(parse_args()))

File: scrape_schema.js
const puppeteer = require('puppeteer');
const fs = require('fs').promises;
const path = require('path');

const BASE_URL = 'https://yellowslate.com';
const DEFAULT_SCRAPED_DATA_DIR = path.join(process.cwd(), 'scraped_data');

async function getAllSchools(outputDir = DEFAULT_SCRAPED_DATA_DIR, cities = null) {
  const schools = [];
  const outputPath = path.resolve(outputDir);
  
  let cityDirs = await fs.readdir(outputPath, { withFileTypes: true });
  cityDirs = cityDirs.filter(d => d.isDirectory() && !d.name.startsWith('_'));
  
  for (const cityDir of cityDirs) {
    const citySlug = cityDir.name;
    if (cities && !cities.includes(citySlug)) continue;
    
    const cityPath = path.join(outputPath, citySlug);
    const entries = await fs.readdir(cityPath, { withFileTypes: true });
    
    for (const entry of entries) {
      if (!entry.isDirectory() || entry.name.startsWith('_')) continue;
      
      const aboutPath = path.join(cityPath, entry.name, 'about.json');
      try {
        const aboutData = JSON.parse(await fs.readFile(aboutPath, 'utf-8'));
        if (aboutData.slug) {
          schools.push({ city: citySlug, folder: entry.name, slug: aboutData.slug, name: aboutData.name });
        }
      } catch (e) {
        // Skip files that can't be read
      }
    }
  }
  return schools;
}

function extractJsonLd(html) {
  const schemaData = [];
  const scriptRegex = /<script[^>]*type="application\/ld\+json"[^>]*>([\s\S]*?)<\/script>/gi;
  let match;
  
  while ((match = scriptRegex.exec(html)) !== null) {
    try {
      const data = JSON.parse(match[1]);
      schemaData.push(data);
    } catch (e) {
      // Skip invalid JSON
    }
  }
  return schemaData;
}

function parseBreadcrumb(data) {
  if (data['@type'] !== 'BreadcrumbList') return null;
  const items = data.itemListElement || [];
  return {
    items: items.map(item => ({
      position: item.position,
      name: item.name,
      url: item.item
    }))
  };
}

function parseLocalBusiness(data) {
  if (data['@type'] !== 'LocalBusiness') return null;
  return {
    name: data.name,
    image: data.image,
    url: data.url,
    telephone: data.telephone,
    address: data.address,
    geo: data.geo
  };
}

function parseReviewSchema(data) {
  if (data['@type'] !== 'Review') return null;
  const itemList = data.itemListElement || [];
  const reviews = itemList.map(review => {
    const itemReviewed = review.itemReviewed || {};
    const author = review.author || {};
    const rating = review.reviewRating || {};
    return {
      school_name: itemReviewed.name,
      school_address: itemReviewed.address,
      school_url: itemReviewed.url,
      aggregate_rating: itemReviewed.aggregateRating,
      author: author.name,
      date_published: review.datePublished,
      rating_value: rating.ratingValue,
      best_rating: rating.bestRating,
      worst_rating: rating.worstRating,
      review_body: review.reviewBody,
      publisher: review.publisher?.name
    };
  });
  return {
    number_of_items: data.numberOfItems,
    reviews
  };
}

function processSchemaData(schemaList) {
  const result = {
    breadcrumbs: [],
    local_business: null,
    reviews: null
  };
  
  for (const schema of schemaList) {
    if (Array.isArray(schema)) {
      for (const item of schema) {
        const breadcrumb = parseBreadcrumb(item);
        if (breadcrumb) result.breadcrumbs = breadcrumb.items;
        
        const localBusiness = parseLocalBusiness(item);
        if (localBusiness) result.local_business = localBusiness;
        
        const reviewSchema = parseReviewSchema(item);
        if (reviewSchema) result.reviews = reviewSchema;
      }
    } else {
      const breadcrumb = parseBreadcrumb(schema);
      if (breadcrumb) result.breadcrumbs = breadcrumb.items;
      
      const localBusiness = parseLocalBusiness(schema);
      if (localBusiness) result.local_business = localBusiness;
      
      const reviewSchema = parseReviewSchema(schema);
      if (reviewSchema) result.reviews = reviewSchema;
    }
  }
  return result;
}

async function scrapeSchoolSchema(browser, school) {
  const { slug, name } = school;
  if (!slug) return null;
  
  const detailUrl = `${BASE_URL}${slug}`;
  const page = await browser.newPage();
  
  try {
    await page.goto(detailUrl, { waitUntil: 'domcontentloaded', timeout: 60000 });
    await page.waitForTimeout(2000);
    
    const html = await page.content();
    const schemaList = extractJsonLd(html);
    
    if (!schemaList.length) {
      console.log(`  No JSON-LD schema found for ${name}`);
      return null;
    }
    
    const processed = processSchemaData(schemaList);
    processed.url = detailUrl;
    processed.scraped_at = new Date().toISOString();
    
    return processed;
  } catch (e) {
    console.log(`  Error scraping ${detailUrl}: ${e.message}`);
    return null;
  } finally {
    await page.close();
  }
}

async function run(args = {}) {
  const outputDir = args.outputDir || DEFAULT_SCRAPED_DATA_DIR;
  const cityFilter = args.city || null;
  const limit = args.limit || null;
  
  console.log('Finding schools to process...');
  const schools = await getAllSchools(outputDir, cityFilter);
  console.log(`Found ${schools.length} schools`);
  
  if (limit) {
    schools.splice(limit);
    console.log(`Limited to ${limit} schools`);
  }
  
  const browser = await puppeteer.launch({ headless: true });
  
  let processed = 0;
  let failed = 0;
  
  for (const school of schools) {
    console.log(`Processing: ${school.city}/${school.folder}`);
    
    const schemaData = await scrapeSchoolSchema(browser, school);
    
    if (schemaData) {
      const schemaPath = path.join(outputDir, school.city, school.folder, 'schema.json');
      await fs.writeFile(schemaPath, JSON.stringify(schemaData, null, 2));
      processed++;
      console.log(`  Saved schema.json`);
    } else {
      failed++;
      console.log(`  Failed to get schema data`);
    }
  }
  
  await browser.close();
  
  console.log('\n' + '='.repeat(60));
  console.log('SCHEMA SCRAPE SUMMARY');
  console.log('='.repeat(60));
  console.log(`Total schools: ${schools.length}`);
  console.log(`Processed: ${processed}`);
  console.log(`Failed: ${failed}`);
  console.log('='.repeat(60));
}

function parseArgs() {
  const args = {
    outputDir: DEFAULT_SCRAPED_DATA_DIR,
    city: null,
    limit: null
  };
  
  for (const arg of process.argv.slice(2)) {
    if (arg.startsWith('--output-dir=')) {
      args.outputDir = arg.split('=')[1];
    } else if (arg.startsWith('--city=')) {
      args.city = arg.split('=')[1].split(',').map(c => c.trim().toLowerCase()).filter(Boolean);
    } else if (arg.startsWith('--limit=')) {
      args.limit = parseInt(arg.split('=')[1], 10) || null;
    }
  }
  
  return args;
}

run(parseArgs()).catch(console.error);